compiler/src/dotty/tools/dotc/typer/RefChecks.scala

```diff
@@ -926,6 +928,81 @@ object RefChecks {
   }
 
   val NoLevelInfo: RefChecks.OptLevelInfo = new OptLevelInfo()
+
+  /**
+   * Verify that references in the user-defined @implicitNotFound message are valid.
+   * (i.e. they refer to a type variable that really occurs in the signature of the annotated symbol.)
+   */
+  private object checkImplicitNotFoundAnnotation:
+
+    /**
+     * Warns if the class or trait has an @implicitNotFound annotation
+     * with invalid type variable references.
+     */
+    def template(sd: SymDenotation)(using Context): Unit =
+      for
+        annotation <- sd.annotations
+        if annotation.symbol == ctx.definitions.ImplicitNotFoundAnnot
+        (arg, l) <- annotation.arguments.collect {
+          case l@Literal(c: Constant) => (c.stringValue, l)
+        }
+      do forEachTypeVariableReferenceIn(arg) { case (ref, start) =>
+        if !sd.typeParams.exists(_.denot.name.show == ref) then
+          reportInvalidReferences(l, ref, start, sd)
+      }
+
+    /**
+     * Warns if the def has parameters with an @implicitNotFound annotation
+     * with invalid type variable references.
+     */
+    def defDef(sd: SymDenotation)(using Context): Unit =
+      for
+        param <- sd.paramSymss.flatten
+        annotation <- param.annotations
+        if annotation.symbol == ctx.definitions.ImplicitNotFoundAnnot
+        (arg, l) <- annotation.arguments.collect {
+          case l@Literal(c: Constant) => (c.stringValue, l)
+        }
+      do forEachTypeVariableReferenceIn(arg) { case (ref, start) =>
+        if !sd.paramSymss.flatten.exists(_.name.show == ref) then
+          reportInvalidReferences(l, ref, start, sd)
+      }
+
+    /** Reports an invalid reference to a type variable <pre>typeRef</pre> that was found in <pre>l</pre> */
+    private def reportInvalidReferences(
+      l: Literal,
+      typeRef: String,
+      offsetInLiteral: Int,
+      sd: SymDenotation
+    )(using Context) =
+      val msg = InvalidReferenceInImplicitNotFoundAnnotation(
+        typeRef, if (sd.isConstructor) "the constructor" else sd.name.show)
+      val span = l.span.shift(offsetInLiteral + 1) // +1 because of 0-based index
+      val pos = ctx.source.atSpan(span.withEnd(span.start))
```

Comment:
You can call `startPos` on a Span to get a Span that starts and end at the start of the original one.